databricks/koalas/datetime.py

```diff
@@ -0,0 +1,154 @@
+#
+# Copyright (C) 2019 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Date/Time related functions on Koalas Series
+"""
+
+import functools
+from pyspark.sql.types import DateType, TimestampType, LongType, StringType
+
+from databricks.koalas.series import Series
+from databricks.koalas.utils import lazy_property
+
+import pyspark.sql.functions as F
+
+
+def defer_to_pandas(output_type):
+    """Wraps a function that operates on pd.Series to work with DatetimeMethods.
+
+    The origin function operates on pd.Series and the wrapped function operates
+    on DatetimeMethods.
+    """
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(self, *args):
+            scol = F.pandas_udf(
+                lambda series: func(series, *args),
+                output_type
+            )(self._data._scol)
+            return Series(
+                scol, anchor=self._data._kdf, index=self._data._index_info)
+        return wrapper
+    return decorator
+
+
+def defer_to_spark(output_type):
+    """Wraps a function that operates on spark.sql.Column to work with DatetimeMethods.
+
+    The origin function operates on spark.sql.Column and the wrapped function
+    operates on DatetimeMethods.
+    """
+    def decorator(func):
```

Comment:
Sorry for coming late to the review, but I would like to make a few improvement before we merge this. This is more complicated than it ought to and it is going to be harder to refactor in the future. We should be able to make the whole body of this function simpler along this sketch:

```py
class DateTimeMethods(object):
  ...
   @lazy_property
    def year(self) -> ks.Series: # Let's get the type hints too.
      return defer_to_spark(self, LongType(), F.year)

def defer_to_spark(kseries, return_type, spark_function):
  scol = func(kseries._data._scol, *args).cast(return_type)
  return Series(scol, anchor=kseries._kdf, index=kseries._index_info)
```

I have not tried this code, but it should be pretty much all you need.

Reply:
I have reimplemented some of the function using pandas_wrap and Series._column_op. 

It looks a little more verbose than before but it doesn't need extra decorator now. If this is what you want I can go ahead and convert the rest of them and remove the two decorators `defer_to_spark` and `defer_to_pandas`.

Reply:
@icexelloss it looks a bit verbose indeed. Since most of the return types are of the type, I think we can refactor it this way, taking your example below:

```py
class DateTimeMethods(object):

    def strftime(self, date_format) -> ks.Series:
        """
        Convert to String Series using specified date_format.
        """
        return self._wrap_string(lamba col: col.dt.strftime(date_format))

    @lazy_property
    def weekofyear(self) -> ks.Series:
        """
        The week ordinal of the year.
        """
        return self._wrap_spark(F.weekofyear)

    def _wrap_string(self, pd_col_fun, return_dtype=StringType()):
       return  pandas_wraps(
            function=lambda x: pd_col_fun(x),
            return_col=return_dtype
        )(self._data)

  def _wrap_spark(self, sp_col_fun, return_dtype=LongType()):
      return _column_op(
            lambda col: sp_col_fun(col).cast(return_dtype)
        )(self._data)
```

What do you think?

Reply:
Yeah I think that looks good to me but I'd suggest removing the default return type for those two wrappers. And I assume `_wrap_string` should be `_wrap_pandas`?

Reply:
Although, then what these two wrappers are really doing is just saving us from writing 
`(self._data)`...

Reply:
Well, that and the `cast` call in `_wrap_spark` case. It looks fine to me either way. Having the wrappers does make the implementation of these functions cleaner so even if it's just saving the `cast` and `(self._data)` so it looks to me to having them. Maybe sth like this?

```
class DateTimeMethods(object):

    def strftime(self, date_format) -> ks.Series:
        """
        Convert to String Series using specified date_format.
        """
        return self._wrap_pandas(lamba col: col.dt.strftime(date_format), StringType())

    @lazy_property
    def weekofyear(self) -> ks.Series:
        """
        The week ordinal of the year.
        """
        return self._wrap_spark(F.weekofyear, LongType())

    def _wrap_pandas(self, pd_col_fun, return_dtype):
       return  pandas_wraps(
            function=lambda x: pd_col_fun(x),
            return_col=return_dtype
        )(self._data)

  def _wrap_spark(self, sp_col_fun, return_dtype):
      return _column_op(
            lambda col: sp_col_fun(col).cast(return_dtype)
        )(self._data)
```